### PR TITLE
Scene conditions

### PR DESCRIPTION
**Pull request by elliotmassen**
This PR adds the ability to specify scene conditions. It is dependent on #235.

### Scene conditions
Conditions can be added to scenes in the following way:

```yaml
conversation:
  id: with_scene_conditions
  scenes:
    opening_scene:
      intents:
        - u:
            i: opening_user_s1
            scene: scene1
        - u:
            i: opening_user_none
        - b: 
            i: opening_bot_response
            completes: true
    scene1:
      conditions:
        - condition:
            operation: is_not_set
            attributes:
              attribute1: user.user_email
      intents:
        - b: 
            i: scene1_bot_response
            completes: true
```

#### Functionality
In terms of functionality, the scene conditions will be pulled from scenes onto the respective cross-scene intents ("said across scenes"). This happens in two places, for opening intents in `EIModelOpeningIntents.collectSceneIntents` and for regular intents in `Intent.getConditionsFromDestinationScene`. In both cases the graph is followed from the cross-scene intent to the participant that listens for it and then up to the scene that the participant is within. From here the conditions are retrieved and added alongside the intent (& potential conversation conditions) for checking.

Previously this PR attempted to use the `ConversationManager` to achieve this. The scene conditions were retrieved and added when calling `userSaysToBotAcrossScenes` or `botSaysToUserAcrossScenes`. While this worked, it meant that the scene conditions were persisted against the intents in the graph, and would be re-added each time, causing consistency issues (as well as memory issues).

#### Behaviour
In the example conversation given above, if the bot receives a 'opening_user_s1' intent, it will have access to the scene1's conditions and these will be checked. If the conditions pass the conversation will progress to scene1, if they fails it will result in a 'NoMatchResponse'. In the future, if we were to allow multiple occurrences of the same intent and we were to have a sequential block of the same intent pointing to various scenes, each destination scenes' conditions would be checked and the conversation would progress to the first scene with passing conditions.

### CODE REVIEW

File: src/ConversationEngine/ConversationStore/EIModels/EIModelOpeningIntents.php [hunk 1]
```diff
@@ -38,27 +41,16 @@ public static function validate(array $response, $additionalParameter = null): b
      * @param array|array $response
      * @param null $additionalParameter
      * @return EIModel
-     * @throws \Illuminate\Contracts\Container\BindingResolutionException
-     * @throws \Exception
+     * @throws BindingResolutionException
+     * @throws Exception
```
Comment (istos) [inline src/ConversationEngine/ConversationStore/EIModels/EIModelOpeningIntents.php]: Is this still required?

File: src/ConversationEngine/ConversationStore/EIModels/EIModelOpeningIntents.php [hunk 2]
```diff
@@ -7,7 +7,10 @@
 use Countable;
 use Ds\Map;
 use Ds\Set;
+use Exception;
```
Comment (istos) [inline src/ConversationEngine/ConversationStore/EIModels/EIModelOpeningIntents.php]: Are we actually using Exceptions?

File: src/Conversation/Intent.php [hunk 3]
```diff
@@ -303,4 +303,48 @@ public function copyNonCoreAttributes(Intent $intent): void
             $this->addAttribute($attribute);
         }
     }
+
+    /**
+     * @return Map
+     */
+    public function getAllConditions(): Map
+    {
+        $intents = new Map();
```
[the remark this answers is not in the document]
Comment (elliotmassen) [reply src/Conversation/Intent.php]: Ah yes, this should've been `$intents` - this is now updated.

File: src/ConversationEngine/ConversationStore/EIModels/EIModelOpeningIntents.php (hunk 1 above)
Comment (elliotmassen) [reply src/ConversationEngine/ConversationStore/EIModels/EIModelOpeningIntents.php]: These were not required, I've now updated the method to only throw a `EIModelCreatorException`.

File: src/ConversationEngine/ConversationStore/EIModels/EIModelOpeningIntents.php (hunk 2 above)
Comment (elliotmassen) [reply src/ConversationEngine/ConversationStore/EIModels/EIModelOpeningIntents.php]: After the recent update we are not. This has now been removed.

**Pull request merged**